File: app/src/main/java/com/animsh/moviem/activities/MovieDetailsActivity.java
package com.animsh.moviem.activities;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.lifecycle.ViewModelProvider;

import com.animsh.moviem.R;
import com.animsh.moviem.databinding.ActivityMovieDetailsBinding;
import com.animsh.moviem.model.Cast;
import com.animsh.moviem.model.Crew;
import com.animsh.moviem.response.moviesresponse.MovieResult;
import com.animsh.moviem.viewmodels.moviesviewmodel.MovieCreditsViewModel;
import com.animsh.moviem.viewmodels.moviesviewmodel.MovieDetailsVieModel;
import com.animsh.moviem.viewmodels.moviesviewmodel.RecommendedMoviesViewModel;
import com.animsh.moviem.viewmodels.moviesviewmodel.SimilarMoviesViewModel;

import java.util.ArrayList;
import java.util.List;

public class MovieDetailsActivity extends AppCompatActivity {

    private ActivityMovieDetailsBinding activityMovieDetailsBinding;
    private MovieDetailsVieModel movieDetailsVieModel;
    private SimilarMoviesViewModel similarMoviesViewModel;
    private RecommendedMoviesViewModel recommendedMoviesViewModel;
    private MovieCreditsViewModel movieCreditsViewModel;

    private List<MovieResult> movieResults = new ArrayList<>();
    private List<Cast> casts = new ArrayList<>();
    private List<Crew> crews = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityMovieDetailsBinding = DataBindingUtil.setContentView(this, R.layout.activity_movie_details);
        movieDetailsVieModel = new ViewModelProvider(this).get(MovieDetailsVieModel.class);
        similarMoviesViewModel = new ViewModelProvider(this).get(SimilarMoviesViewModel.class);
        recommendedMoviesViewModel = new ViewModelProvider(this).get(RecommendedMoviesViewModel.class);
        movieCreditsViewModel = new ViewModelProvider(this).get(MovieCreditsViewModel.class);
        int movieId = getIntent().getIntExtra("movieId", -1);
        getMovieDetails(movieId);

        setSupportActionBar(activityMovieDetailsBinding.toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        activityMovieDetailsBinding.tabLayout.setupWithViewPager(activityMovieDetailsBinding.viewPager);
        activityMovieDetailsBinding.viewPager.setOffscreenPageLimit(3);
        ViewPagerAdapter viewPagerAdapter = new ViewPagerAdapter(getSupportFragmentManager(), 0);
        viewPagerAdapter.addFragment(new SimilarMoviesFragment(movieResults, movieId, 1), "MORE LIKE THIS");
        viewPagerAdapter.addFragment(new SimilarMoviesFragment(movieResults, movieId, 2), "RECOMMENDATIONS");
        viewPagerAdapter.addFragment(new CreditsFragment(crews, casts, movieId), "CAST & CREW");
        activityMovieDetailsBinding.viewPager.setAdapter(viewPagerAdapter);
    }


    void getMovieDetails(int movieId) {
        movieDetailsVieModel.getMovieDetails(movieId, getString(R.string.api_key)).observe(this, movieDetailResponse -> {
            activityMovieDetailsBinding.setMovie(movieDetailResponse);
        });
    }

    private class ViewPagerAdapter extends FragmentPagerAdapter {
        List<Fragment> fragmentList = new ArrayList<>();
        List<String> titleList = new ArrayList<>();

        public ViewPagerAdapter(@NonNull FragmentManager fm, int behavior) {
            super(fm, behavior);
        }

        public void addFragment(Fragment fragment, String title) {
            fragmentList.add(fragment);
            titleList.add(title);
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            return fragmentList.get(position);
        }

        @Override
        public int getCount() {
            return fragmentList.size();
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            return titleList.get(position);
        }
    }
}

File: app/src/main/java/com/animsh/moviem/ui/home/movies/details/MovieDetailsActivity.kt
package com.animsh.moviem.ui.home.movies.details

import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.res.ResourcesCompat
import androidx.lifecycle.ViewModelProvider
import com.animsh.moviem.R
import com.animsh.moviem.adapters.ViewPagerAdapter
import com.animsh.moviem.data.database.entity.FavoriteMovieEntity
import com.animsh.moviem.data.viewmodels.MoviesViewModel
import com.animsh.moviem.databinding.ActivityMovieDetailsBinding
import com.animsh.moviem.models.movie.Movie
import com.animsh.moviem.ui.home.movies.MoviesBottomSheet
import com.animsh.moviem.util.Constants
import com.google.android.material.button.MaterialButton
import com.google.android.material.snackbar.Snackbar
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class MovieDetailsActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMovieDetailsBinding
    private lateinit var moviesViewModel: MoviesViewModel
    private var movieSaved: Boolean = false
    private var savedMovieId: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMovieDetailsBinding.inflate(layoutInflater)
        setContentView(binding.root)
        moviesViewModel = ViewModelProvider(this).get(MoviesViewModel::class.java)

        binding.apply {
            val data: Movie? = intent.getParcelableExtra<Movie>("movie")
            movie = data
            tabLayout.setupWithViewPager(viewPager)
            viewPager.offscreenPageLimit = 3
            val viewPagerAdapter = ViewPagerAdapter(supportFragmentManager, 0)
            viewPagerAdapter.addFragment(SRMoviesFragment(data!!.id, 0), "More Like This")
            viewPagerAdapter.addFragment(SRMoviesFragment(data.id, 1), "Recommendations")
            viewPagerAdapter.addFragment(MovieCreditsFragment(data.id), "Crew & Cast")
            viewPager.adapter = viewPagerAdapter

            shareBtn.setOnClickListener {
                Picasso.get()
                    .load(Constants.IMAGE_W500 + data.posterPath)
                    .into(object : Target {
                        override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
                            val intent = Intent(Intent.ACTION_SEND)
                            intent.type = "image/*"
                            intent.putExtra(
                                Intent.EXTRA_STREAM,
                                Constants.getLocalBitmapUri(
                                    bitmap!!,
                                    this@MovieDetailsActivity,
                                    data.title!!
                                )
                            )
                            val dataText = "${data.title}\n${data.homepage!!}"
                            intent.putExtra(
                                Intent.EXTRA_TEXT,
                                dataText
                            )
                            startActivity(Intent.createChooser(intent, "Share Movie"))
                        }

                        override fun onBitmapFailed(e: Exception?, errorDrawable: Drawable?) {
                            Log.d(MoviesBottomSheet.TAG, "onBitmapFailed: " + e?.message)
                            val intent = Intent(Intent.ACTION_SEND)
                            intent.type = "text/plain"
                            val dataText = "${data.title}\n${data.homepage!!}"
                            intent.putExtra(
                                Intent.EXTRA_TEXT,
                                dataText
                            )
                            startActivity(Intent.createChooser(intent, "Share Movie"))
                        }

                        override fun onPrepareLoad(placeHolderDrawable: Drawable?) {}
                    })
            }

            addToMyList.setOnClickListener {
                if (movieSaved) {
                    removeFromFav(addToMyList)
                } else {
                    saveToFav(addToMyList)
                }
            }

            checkFavMovieStatus()
        }
    }

    private fun saveToFav(addToMyList: MaterialButton) {
        val favoriteMovieEntity = FavoriteMovieEntity(0, binding.movie!!)
        moviesViewModel.insertFavMovie(favoriteMovieEntity)
        addToMyList.icon = ResourcesCompat.getDrawable(resources, R.drawable.ic_check, null)
        showMessage("added to my list!!")
        movieSaved = true
    }

    private fun removeFromFav(addToMyList: MaterialButton) {
        val favoriteMovieEntity = FavoriteMovieEntity(savedMovieId, binding.movie!!)
        moviesViewModel.deleteFavMovie(favoriteMovieEntity)
        addToMyList.icon = ResourcesCompat.getDrawable(resources, R.drawable.ic_add, null)
        showMessage("removed from my list!!")
        movieSaved = false
    }

    private fun checkFavMovieStatus() {
        moviesViewModel.readFavMovie.observe(this, { favoriteEntity ->
            try {
                for (savedMovie in favoriteEntity) {
                    if (savedMovie.result.id == binding.movie!!.id) {
                        binding.addToMyList.icon =
                            ResourcesCompat.getDrawable(resources, R.drawable.ic_check, null)
                        movieSaved = true
                        savedMovieId = savedMovie.id
                    }
                }
            } catch (e: Exception) {
                Log.d("TAGTAGTAG", "checkFavMovieStatus: " + e.message)
            }
        })
    }

    private fun showMessage(message: String) {
        Snackbar.make(
            binding.layoutMovieDetails,
            message,
            Snackbar.LENGTH_SHORT
        ).setAction("Okay") {}
            .show()
    }

}